### PR TITLE
Allow render HTML with css/js external links

### PR DESCRIPTION
**Pull request by lunny**
This PR allows to render some HTMLs which has `<style>`, `<script>` tags when `SANITIZER` disabled. This also makes it simpler when configuring some external renderers.

### CODE REVIEW

**Review by techknowlogick — CHANGES_REQUESTED** [review]
I'm concerned with this PR because someone could link to `/user/repo/render/....` and have XSS execute. I know that they chose to do this by disabling SANITIZER, but this seems especially dangerous. If merged, we should at least create a big warning in docs, and perhaps in admin saying the config is insecure.

**Comment by lunny** [comment]
> I'm concerned with this PR because someone could link to `/user/repo/render/....` and have XSS execute. I know that they chose to do this by disabling SANITIZER, but this seems especially dangerous. If merged, we should at least create a big warning in docs, and perhaps in admin saying the config is insecure.

Warning added in the docs.

**Comment by techknowlogick** [comment]
Could we use Content Security Policy headers to prevent any XSS?

File: modules/markup/renderer.go
Comment (delvh) [inline modules/markup/renderer.go]: `scrolling=no` is apparently deprecated and should be replaced (or accompanied) by `style="overflow: hidden"`, which however does not appear to be supported on all browsers.
Should we supply both (or add an additional style class)?

File: routers/web/repo/render.go
Comment (delvh) [inline routers/web/repo/render.go]:
```suggestion
			ctx.NotFound("GetBlobByPath", err)
```

File: routers/web/repo/render.go
Comment (delvh) [inline routers/web/repo/render.go]:
```suggestion
	if ctx.Repo.TreePath != "" {
```

File: routers/web/repo/view.go
Comment (delvh) [inline routers/web/repo/view.go:360]: Why was this file path changed?

File: routers/web/repo/view.go
Comment (delvh) [inline routers/web/repo/view.go:635]: And again, the call was changed?

File: routers/web/repo/render.go
Comment (delvh) [inline routers/web/repo/render.go]:
```suggestion
```

File: routers/web/repo/view.go
Comment (lunny) [reply routers/web/repo/view.go:360]: Because we need the relative path of the file but not only the filename. Sometimes the file will be put into a sub directory of git tree.

File: routers/web/repo/view.go
Comment (lunny) [reply routers/web/repo/view.go:635]: As above.

**Review by delvh — APPROVED** [review]
Apart from the deprecation issue where we have yet to find a solution LGTM.

File: modules/markup/renderer.go
Comment (wxiaoguang) [inline modules/markup/renderer.go]: AppURL => AppSubURL?
user => escaping
repo => escaping
BranchNameSubURL: (it should have been escaped)
name="ifd" ... ifd.document: not sure it follows the HTML/JS standard to use the `name` as a variable name.
`/>` => `></iframe>` IIRC for standard

File: routers/web/repo/render.go
Comment (wxiaoguang) [inline routers/web/repo/render.go]: Does EscapeControlReader really work here? EscapeControlReader requires some CSS and JS to work IIRC.

File: modules/markup/renderer.go
Comment (lunny) [reply modules/markup/renderer.go]: I will add `style="overflow: hidden"`, split style to a file looks like not a good idea because it's a embed iframe.

**Comment by lunny** [comment]
@delvh @wxiaoguang both done.

File: modules/markup/console/console.go
Comment (wxiaoguang) [inline modules/markup/console/console.go]: Comment? `sanitize`?

**Review by wxiaoguang — COMMENTED** [review]
The introduced `DisplayInIFrame` seems redundancy and confusing. Many render doesn't support the `DisplayInIFrame`, and it's only used for the external render.

In my mind, the logic might be:

```
if (externalRender) {
    if (renderInIframe) {
        show iframe
    } else {
        render in page
    }
} else {
    render in page
}
```

And the `DISABLE_SANITIZER` + `USE_IFRAME` makes it difficult to understand all combinations.
* DISABLE_SANITIZER=true, USE_IFRAME=true
* DISABLE_SANITIZER=false, USE_IFRAME=true
What's the difference of them? Does USE_IFRAME require DISABLE_SANITIZER=true?

**Comment by lunny** [comment]
> The introduced `DisplayInIFrame` seems redundancy and confusing. Many render doesn't support the `DisplayInIFrame`, and it's only used for the external render.
> 
> In my mind, the logic might be:
> 
> ```
> if (externalRender) {
>     if (renderInIframe) {
>         show iframe
>     } else {
>         render in page
>     }
> } else {
>     render in page
> }
> ```
> 
> And the `DISABLE_SANITIZER` + `USE_IFRAME` makes it difficult to understand all combinations.
> 
> * DISABLE_SANITIZER=true, USE_IFRAME=true
> * DISABLE_SANITIZER=false, USE_IFRAME=true
>   What's the difference of them? Does USE_IFRAME require DISABLE_SANITIZER=true?

DISABLE_SANITIZER is just available when USE_IFRAME is false. I have changed the logic.

File: modules/markup/external/external.go
Comment (delvh) [inline modules/markup/external/external.go]: That sounds more wrong than previously...
I'm in favor of the previous version.
Otherwise we will at somepoint get hard to debug problems.

Comment (wxiaoguang) [reply modules/markup/external/external.go]: Since 1.17 hasn't been released, how about introducing a setting option `RENDER_PAGE_MODE=default/no-sanitizer/iframe` instead? To replace both DISABLE_SANITIZER+USE_IFRAME.


`RENDER_PAGE_MODE=default`: default mode as 1.16, use sanitizer
`RENDER_PAGE_MODE=no-sanitizer`: do not use sanitizer, render in the page directly
`RENDER_PAGE_MODE=iframe`: render in the iframe without sanitizer

Comment (wxiaoguang) [reply modules/markup/external/external.go]: ~~I read the code again, `renderFile` is only called once in `renderCode`. So the `DisplayInIFrame` doesn't need to be a member of the `Render`.~~ hmm, some misunderstanding .

Maybe `renderIFrame` could follow the config `RENDER_PAGE_MODE` , output the `<iframe>` directly if the iframe is used, instead of making the render output the `<iframe>` (just some guess, not sure whether feasible)

Comment (wxiaoguang) [reply modules/markup/external/external.go]: I have some thoughts about the refactoring of MarkupRender system.

First, the Render interface should be splitted to InternalRender and ExternalRender, then no confusing `SanitizerDisabled/DisplayInIFrame` function for internal renders.

Then, in the `renderFile`, it could do type-cast to check if the matched render is an external render, if yes, check if it will do iframe rendering, if it will do, then output the iframe.

If you do not mind, I could try to propose a PR (today, to make sure no delay for 1.17 milestone)

Comment (lunny) [reply modules/markup/external/external.go]: reverted.

File: modules/markup/renderer.go
Comment (wxiaoguang) [inline modules/markup/renderer.go:290]: One more question , if `renderFile` is called somewhere (a normal page, not the `/render/*`) and it matches an external render with `AllowIFrame=false, DisplayInIFrame=true`, will the content be rendered directly in current page? If yes, is it allowed or should it be forbidden?

Comment (lunny) [reply modules/markup/renderer.go:290]: In fact, `AllowIFrame=false, DisplayInIFrame=true` is used by the link in the iframe.

Comment (wxiaoguang) [reply modules/markup/renderer.go:290]: I mean "(a normal page, not the /render/*)", is it guaranteed that no other code would call `renderFile` with "AllowIFrame=false, DisplayInIFrame=true" ? If it happens, the content be rendered directly in normal page.

File: modules/markup/renderer.go
Comment (techknowlogick) [inline modules/markup/renderer.go]: can you add the sandbox attribute (more details here: https://www.w3schools.com/tags/att_iframe_sandbox.asp). The attribute will prevent JS inside iframe from writing to parent window.

Comment (lunny) [reply modules/markup/renderer.go]: Done.

**Review by techknowlogick — APPROVED** [review]

File: modules/markup/renderer.go
Comment (delvh) [inline modules/markup/renderer.go]: Doesn't `allow-scripts` defeat the whole point of trying to prevent XSS attacks?

**Comment by wxiaoguang** [comment]
Personally I do not think this design is good. It would make the  config system and Render interface confusing.

If most people like, I have no objection.

**Comment by delvh** [comment]
You are referring to https://github.com/go-gitea/gitea/pull/19017#discussion_r895168355, correct?
I am generally in favor of it, the only problem I can foresee is that it would be a breaking change, correct?
Or is the `DISABLE_SANITIZER` not part of 1.16?

**Comment by lunny** [comment]
> You are referring to [#19017 (comment)](https://github.com/go-gitea/gitea/pull/19017#discussion_r895168355), right? I am generally in favor of it, the only problem I can foresee is that it would be a breaking change, correct? Or is the `DISABLE_SANITIZER` configuration not part of 1.16?

I think this is not a break change. This PR allows output a throughout `HTML` document with scripts/css but not only a `DIV`. Before this PR, external tools have to remove `HTML` tag and `HEAD` tag.

File: modules/markup/renderer.go
Comment (techknowlogick) [reply modules/markup/renderer.go]: IIRC allow-scripts allows JS to be run inside the iframe, but doesn't allow access to the parent. (this is only an assumption though based on my reading of MDN, I'd be very happy to be corrected).

Comment (delvh) [reply modules/markup/renderer.go]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#scripting :
> From the inside of a frame, a script can get a reference to its parent window with [window.parent](https://developer.mozilla.org/en-US/docs/Web/API/Window/parent).

The closest thing I found regarding your point:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox :
> allow-scripts: Lets the resource run scripts (but not create popup windows).

This sounds to me like the only thing blocked are pop-ups (`window.alert`, `window.confirm`, ...), but nothing else...

**Comment by wxiaoguang** [comment]
I propose a PR 
* https://github.com/lunny/gitea/pull/23

Major changes:
1. Add missing `USE_IFRAME` to `app.ini`
2. Clarify how `USE_IFRAME` works: it suppresses DISABLE_SANITIZER option and there will be no sanitizer.
3. Introduce `markup.PostProcessRenderer` and `markup.ExternalRenderer`, only the external render needs to implement the `markup.ExternalRenderer` interface (aka `SanitizerDisabled` and `DisplayInIFrame`)
4. Introduce `ctx.InStandalonePage`, it's simpler than `AllowIFrame`, only the standalone render page needs `InStandalonePage=true`, all other calls to renders just leave it to false and the `iframe` will be generated accordingly.

**Comment by wxiaoguang** [comment]
Hmm, still some nits should be fixed

The `height` + `overflow: hidden` would make the one-line content not visible.

image

File: modules/markup/renderer.go
Comment (wxiaoguang) [inline modules/markup/renderer.go]:
```suggestion
<iframe src="%s/%s/%s/render/%s/%s"
```

AppSubURL doesn't have the trailing slash, so here the slash is needed.

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Final question, keep `DISABLE_SANITIZER`+`USE_IFRAME`, or use `RENDER_PAGE_MODE=default/no-sanitizer/iframe` instead?

**Comment by lunny** [comment]
> Final question, keep `DISABLE_SANITIZER`+`USE_IFRAME`, or use one option like `RENDER_PAGE_MODE=default/no-sanitizer/iframe` instead?

I vote both.

**Comment by delvh** [comment]
As it is non-breaking, I'm for the enum instead of the booleans approach.

**Comment by lunny** [comment]
> As it is non-breaking, I'm for the enum instead of the booleans approach.

That will make it break imo. `DISABLE_SANITIZER` will be dropped.

**Comment by lunny** [comment]
> [#19017 (comment)](https://github.com/go-gitea/gitea/pull/19017#issuecomment-1153994596):
> 
> > `DISABLE_SANITIZER` is not in 1.16. It's a good chance to make it clear.
> 
> So it is not breaking for users who are on a stable version, which should be almost everyone.

If that, we need a migration and deprecated warning in setting.

**Comment by wxiaoguang** [comment]
Beside the option, still one more thing (sorry for bothering ...)

Now the iframe is using `sandbox="allow-same-origin allow-scripts"`, then the rendered JS can access to the `window.config` which contains CSRF token, then the JS can make requests as the current users. Is it fine?

image

image

**Comment by delvh** [comment]
https://github.com/go-gitea/gitea/pull/19017#discussion_r896029490:
Now we certainly know that we cannot allow that.

If we allow that, we have an obvious and unintended privilege escalation.
I am strongly against allowing the scripts now.

**Comment by wxiaoguang** [comment]
> [#19017 (comment)](https://github.com/go-gitea/gitea/pull/19017#discussion_r896029490): Now we certainly know that we cannot allow that.
> 
> If we allow that, we have an obvious and unintended privilege escalation. I am strongly against allowing the scripts now.

Only `allow-scripts` can be used here. I am working on it.

**Review by techknowlogick — CHANGES_REQUESTED** [review]
temporarily blocking per my discussion w/ delvh

**Comment by wxiaoguang** [comment]
OK, all fixed. Now the iframe is protected by browser's same-origin policy. @techknowlogick 

`Uncaught DOMException: Blocked a frame with origin "null" from accessing a cross-origin frame.`

Refer: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox

If this token is not used, the resource is treated as being from a special origin that always fails the [same-origin policy](https://developer.mozilla.org/en-US/docs/Glossary/Same-origin_policy) (potentially preventing access to [data storage/cookies](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#cross-origin_data_storage_access) and some JavaScript APIs).

File: custom/conf/app.example.ini
Comment (lunny) [inline custom/conf/app.example.ini]: Allowed rules could be defined with `[markup.sanitizer.*]`.

Comment (wxiaoguang) [reply custom/conf/app.example.ini]:
```suggestion
;; * sanitized: Sanitize the content and render it inside current page, default to only allow a few HTML tags and attributes. Customized sanitizer rules can be defined in [markup.sanitizer.*] .
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md
Comment (wxiaoguang) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
  - sanitized: Sanitize the content and render it inside current page, default to only allow a few HTML tags and attributes. Customized sanitizer rules can be defined in `[markup.sanitizer.*]`.
```

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**
